pkg/mutation/system.go

```diff
@@ -98,14 +106,72 @@ func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) (b
 				return false, nil
 			}
 			if cmp.Equal(original, obj) {
+				if *MutationLoggingEnabled {
+					logAppliedMutations("Oscillating mutation.", original, allAppliedMutations)
+				}
 				return false, fmt.Errorf("oscillating mutation for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
 			}
+			if *MutationLoggingEnabled {
+				logAppliedMutations("Mutation applied.", original, allAppliedMutations)
+			}
 			return true, nil
 		}
+		if *MutationLoggingEnabled {
+			allAppliedMutations = append(allAppliedMutations, appliedMutations)
+		}
+	}
+	if *MutationLoggingEnabled {
+		logAppliedMutations("Mutation not converging.", original, allAppliedMutations)
 	}
 	return false, fmt.Errorf("mutation not converging for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
 }
 
+func logAppliedMutations(message string, obj *unstructured.Unstructured, allAppliedMutations [][]types.Mutator) {
+	iterations := []interface{}{}
+	for i, appliedMustations := range allAppliedMutations {
+		if len(appliedMustations) == 0 {
+			continue
+		}
+		var appliedMutationsText strings.Builder
+		mutationSeparator := ""
+		for _, mutator := range appliedMustations {
+			fmt.Fprintf(&appliedMutationsText, "%s%s", mutationSeparator, mutatorKey(mutator))
+			mutationSeparator = ", "
+		}
+		iterations = append(append(iterations, fmt.Sprintf("iteration_%d", i)), appliedMutationsText.String())
+	}
+	if len(iterations) > 0 {
+		logDetails := []interface{}{}
+		logDetails = append(append(logDetails, logging.EventType), logging.MutationApplied)
+		logDetails = append(append(logDetails, logging.ResourceGroup), obj.GroupVersionKind().Group)
+		logDetails = append(append(logDetails, logging.ResourceKind), obj.GroupVersionKind().Kind)
+		logDetails = append(append(logDetails, logging.ResourceAPIVersion), obj.GroupVersionKind().Version)
+		logDetails = append(append(logDetails, logging.ResourceNamespace), obj.GetNamespace())
+		logDetails = append(append(logDetails, logging.ResourceName), obj.GetName())
+		logDetails = append(logDetails, iterations...)
+		log.Info(message, logDetails...)
+	}
+}
+
+func mutatorKey(mutator types.Mutator) string {
+	mutatorType := ""
+	resourceVersion := ""
+	switch castMutator := mutator.(type) {
+	case *AssignMutator:
+		mutatorType = "Assign"
+		resourceVersion = castMutator.assign.GetResourceVersion()
+	case *AssignMetadataMutator:
```

Comment:
nit: Should the default case return something completely different?

Reply:
We can not have a mutator of type different than AssignMutator or AssignMetadataMutator without someone adding it in the code. If you want I can add a default with type "unknown", but this would be dead code.

Reply:
Code removed after latest changes